### ISSUE TITLE
import fails with undefined symbol: _ZTIN7GenICam16GenericExceptionE

### ISSUE THREAD

**Issue by mabl** [srgblnch/python-pylon#3]
Hi,

when using the master branch, the import call fails for me:

```
[mabl@workaholix ~]$ source ~/PhD/Software/python-pylon/env.sh
Environment for Pylon 2 arch64
[mabl@workaholix ~]$ export PYTHONPATH=~/PhD/Software/python-pylon/build/lib.linux-x86_64-2.7/
[mabl@workaholix ~]$ python2
Python 2.7.11 (default, Dec  6 2015, 15:43:46) 
[GCC 5.2.0] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> import pylon
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ImportError: /home/mabl/PhD/Software/python-pylon/build/lib.linux-x86_64-2.7/pylon.so: undefined symbol: _ZTIN7GenICam16GenericExceptionE
>>> 

```

Am I doing anything wrong?

**Comment by mabl** [comment]
So poking blindly, I've tried building this library against pylon 5. Based on the makefile examples I've modified the setup.sh to the following:

```
#source env.sh

echo $LD_LIBRARY_PATH

# Build tools and flags
export LD=${CXX}
export CPPFLAGS=`${PYLON_ROOT}/bin/pylon-config --cflags`
#export CXXFLAGS= #e.g., CXXFLAGS=-g -O0 for debugging
export LDFLAGS=`${PYLON_ROOT}/bin/pylon-config --libs-rpath`
export LDLIBS=`${PYLON_ROOT}/bin/pylon-config --libs`

export CC='g++'

python2 setup.py build
```

This sucessfully builds for me:

```
$ source /opt/pylon5/bin/pylon-setup-env.sh /opt/pylon5

$ ./setup.sh clean
Proceeding with clean operation.

$ ./setup.sh      

running build
running build_ext
cythoning pylon/__init__.pyx to pylon/__init__.cpp
building 'pylon' extension
creating build
creating build/temp.linux-x86_64-2.7
creating build/temp.linux-x86_64-2.7/pylon
g++ -fno-strict-aliasing -march=x86-64 -mtune=generic -O2 -pipe -fstack-protector-strong -DNDEBUG -march=x86-64 -mtune=generic -O2 -pipe -fstack-protector-strong -I/opt/pylon5/include -fPIC -I/usr/include/python2.7 -c pylon/__init__.cpp -o build/temp.linux-x86_64-2.7/pylon/__init__.o
g++ -fno-strict-aliasing -march=x86-64 -mtune=generic -O2 -pipe -fstack-protector-strong -DNDEBUG -march=x86-64 -mtune=generic -O2 -pipe -fstack-protector-strong -I/opt/pylon5/include -fPIC -I/usr/include/python2.7 -c pylon/Logger.cpp -o build/temp.linux-x86_64-2.7/pylon/Logger.o
g++ -fno-strict-aliasing -march=x86-64 -mtune=generic -O2 -pipe -fstack-protector-strong -DNDEBUG -march=x86-64 -mtune=generic -O2 -pipe -fstack-protector-strong -I/opt/pylon5/include -fPIC -I/usr/include/python2.7 -c pylon/Factory.cpp -o build/temp.linux-x86_64-2.7/pylon/Factory.o
g++ -fno-strict-aliasing -march=x86-64 -mtune=generic -O2 -pipe -fstack-protector-strong -DNDEBUG -march=x86-64 -mtune=generic -O2 -pipe -fstack-protector-strong -I/opt/pylon5/include -fPIC -I/usr/include/python2.7 -c pylon/DevInfo.cpp -o build/temp.linux-x86_64-2.7/pylon/DevInfo.o
g++ -fno-strict-aliasing -march=x86-64 -mtune=generic -O2 -pipe -fstack-protector-strong -DNDEBUG -march=x86-64 -mtune=generic -O2 -pipe -fstack-protector-strong -I/opt/pylon5/include -fPIC -I/usr/include/python2.7 -c pylon/Camera.cpp -o build/temp.linux-x86_64-2.7/pylon/Camera.o
creating build/lib.linux-x86_64-2.7
g++ -pthread -shared -Wl,-O1,--sort-common,--as-needed,-z,relro -Wl,--enable-new-dtags -Wl,-rpath,/opt/pylon5/lib64 -I/opt/pylon5/include build/temp.linux-x86_64-2.7/pylon/__init__.o build/temp.linux-x86_64-2.7/pylon/Logger.o build/temp.linux-x86_64-2.7/pylon/Factory.o build/temp.linux-x86_64-2.7/pylon/DevInfo.o build/temp.linux-x86_64-2.7/pylon/Camera.o -L/usr/lib -lpython2.7 -o build/lib.linux-x86_64-2.7/pylon.so
```

When importing that file however, I still get a comparable error:

```
python2                                              
Python 2.7.11 (default, Dec  6 2015, 15:43:46) 
[GCC 5.2.0] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> import pylon
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ImportError: ./pylon.so: undefined symbol: _ZTIN29GenICam_3_0_Basler_pylon_v5_016GenericExceptionE
>>> 
```

**Comment by mabl** [comment]
Hmm hard-coding in setup.py works at least:

```
pylonExtension = Extension('pylon',['pylon/__init__.pyx',
                                    'pylon/Logger.cpp',
                                    'pylon/Factory.cpp',
                                    'pylon/DevInfo.cpp',
                                    'pylon/Camera.cpp'],
                           language="c++",
                           include_dirs=["/opt/pylon5/include", ],
                           library_dirs=["/opt/pylon5/lib64", ],
                           libraries=["pylonbase", "pylonutility", "GenApi_gcc_v3_0_Basler_pylon_v5_0", "GCBase_gcc_v3_0_Basler_pylon_v5_0"],
                           #extra_compile_args=["-std=c++11"]
                           )
```

```
running build
running build_ext
cythoning pylon/__init__.pyx to pylon/__init__.cpp
building 'pylon' extension
creating build
creating build/temp.linux-x86_64-2.7
creating build/temp.linux-x86_64-2.7/pylon
gcc -pthread -fno-strict-aliasing -march=x86-64 -mtune=generic -O2 -pipe -fstack-protector-strong -DNDEBUG -march=x86-64 -mtune=generic -O2 -pipe -fstack-protector-strong -fPIC -I/opt/pylon5/include -I/usr/include/python2.7 -c pylon/__init__.cpp -o build/temp.linux-x86_64-2.7/pylon/__init__.o
gcc -pthread -fno-strict-aliasing -march=x86-64 -mtune=generic -O2 -pipe -fstack-protector-strong -DNDEBUG -march=x86-64 -mtune=generic -O2 -pipe -fstack-protector-strong -fPIC -I/opt/pylon5/include -I/usr/include/python2.7 -c pylon/Logger.cpp -o build/temp.linux-x86_64-2.7/pylon/Logger.o
gcc -pthread -fno-strict-aliasing -march=x86-64 -mtune=generic -O2 -pipe -fstack-protector-strong -DNDEBUG -march=x86-64 -mtune=generic -O2 -pipe -fstack-protector-strong -fPIC -I/opt/pylon5/include -I/usr/include/python2.7 -c pylon/Factory.cpp -o build/temp.linux-x86_64-2.7/pylon/Factory.o
gcc -pthread -fno-strict-aliasing -march=x86-64 -mtune=generic -O2 -pipe -fstack-protector-strong -DNDEBUG -march=x86-64 -mtune=generic -O2 -pipe -fstack-protector-strong -fPIC -I/opt/pylon5/include -I/usr/include/python2.7 -c pylon/DevInfo.cpp -o build/temp.linux-x86_64-2.7/pylon/DevInfo.o
gcc -pthread -fno-strict-aliasing -march=x86-64 -mtune=generic -O2 -pipe -fstack-protector-strong -DNDEBUG -march=x86-64 -mtune=generic -O2 -pipe -fstack-protector-strong -fPIC -I/opt/pylon5/include -I/usr/include/python2.7 -c pylon/Camera.cpp -o build/temp.linux-x86_64-2.7/pylon/Camera.o
creating build/lib.linux-x86_64-2.7
g++ -pthread -shared -Wl,-O1,--sort-common,--as-needed,-z,relro build/temp.linux-x86_64-2.7/pylon/__init__.o build/temp.linux-x86_64-2.7/pylon/Logger.o build/temp.linux-x86_64-2.7/pylon/Factory.o build/temp.linux-x86_64-2.7/pylon/DevInfo.o build/temp.linux-x86_64-2.7/pylon/Camera.o -L/opt/pylon5/lib64 -L/usr/lib -lpylonbase -lpylonutility -lGenApi_gcc_v3_0_Basler_pylon_v5_0 -lGCBase_gcc_v3_0_Basler_pylon_v5_0 -lpython2.7 -o build/lib.linux-x86_64-2.7/pylon.so
```

```
python2
Python 2.7.11 (default, Dec  6 2015, 15:43:46) 
[GCC 5.2.0] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> import pylon
>>> dir(pylon)
['Camera', 'Factory', 'Logger', '_BUILD_VERSION', '_Lock', '_MAINTENANCE_VERSION', '_MAJOR_VERSION', '_MINOR_VERSION', '__DeviceInformation', '__builtins__', '__doc__', '__file__', '__name__', '__package__', '__test__', '_currentThread', '_datetime', '_logger_DEBUG', '_logger_ERROR', '_logger_INFO', '_logger_WARNING', '_weakref', 'lock', 'version_pylon_API', 'version_pylon_API_string', 'version_python_pylon', 'version_python_pylon_string']

>>> pylon.version_pylon_API_string()
'5.0.1-6388'
```

However, I needed to add the calls to Pylon::PylonInitialize and PylonTerminate to ./pylon/Factory.cpp.

Now, If I only had a GigE camera instead of an USB one...

**Comment by srgblnch** [comment]
In the first comment in the issue, looks like the export of the pythonpath overwrites the one made in the env.sh script.

About the next comments, I didn't check with pylon5. In fact, I've only saw this major release for windows when I've started. I've just download it from linux now and the first thing I've saw is the genicam root have changed. I'll try to fix it in the env.sh, I think your proposal in the setup.py will not allow to use any archived version (and by now I need them).

/Sergi.
